docs/backend.md

```diff
@@ -0,0 +1,284 @@
+
+class ResponseType(BaseModel):
+    text: str
+
+async def call_worker():
+    # Get the runtime
+    runtime = await get_runtime()
+
+    # Get a client to the worker endpoint
+    client = await runtime.namespace("your_namespace").component("YourWorker").endpoint("generate").client()
+
+    # Create a request
+    request = RequestType(text="Hello, Dynamo!")
+
+    # Call the dynamo endpoint exposed by the worker
+    responses = await client.generate(request)
+    for response in responses:
+        print(f"Response: {response.text}")
+
+if __name__ == "__main__":
+    asyncio.run(call_worker())
+```
+
+If putting a worker defined to handle OpenAI objects like ChatCompletions
+directly behind an OpenAI `http` service via `llmctl`, you could instead use
+an OpenAI-based client (or `curl`) that communicates with the OpenAI HTTP Service
+and internally routes the requests to the worker(s) instead.
+
+In more advanced scenarios where your worker may operate on some other intermediate format
+that may not directly match an OpenAI-like format, you could setup a separate processor worker
+that does something like the following:
+- Take in OpenAI Chat Completions requests from the HTTP service
+- Convert requests from Chat Completions format to the RequestType format your worker expects
+- Forward requests to the worker(s)
+- Convert responses from the worker's ResponseType back into Chat Completions response format
+- Forward responses back to client
+
+This advanced scenario of a separate OpenAI Processor worker is demonstrated in this
+[vLLM example](examples/llm/README.md).
+
+For a more minimal example of deploying a pipeline of components with a custom
+API that your client can communicate with, see the
+[Hello World example](examples/hello_world/README.md).
+
+## Advanced Features
+
+### KV Routing for LLMs
+
+KV-aware routing is a powerful feature of Dynamo that optimizes for routing
+requests to specific workers while minimizing a specific KV-cache based cost function.
+
+In its simplest form, all a worker needs to do to enable KV-aware routing is to
+publish KV metrics for Dynamo's KV Router to consume:
+```python
+from dynamo.llm import KvMetricsPublisher
+
+class YourWorker:
+    def __init__(self):
+        # Initialize metrics publisher from Dynamo
+        self.metrics_publisher = KvMetricsPublisher()
+
+        # (Optional) Initialize some metrics for the worker/class to track
+        self.request_active_slots = 0
+
+        ###
+        ###  TODO: Verify this code, see if async_init/async_on_start needed
+        ###
+
+        # Send some dummy metrics at initialization time
+        self.metrics_publisher.publish(
+            0,     # request_active_slots
+            1024,  # request_total_slots
+            0,     # kv_active_blocks
+            1024,  # kv_total_blocks
+            0,     # num_requests_waiting
+            0.0,   # gpu_cache_usage_perc
+            0.0,   # gpu_prefix_cache_hit_rate
+        )
```

Comment:
@alec-flowers @tanmayv25 @GuanLuo both vllm/trtllm examples have something like that send some dummy metrics at init time. Is this just so that the KV Router has some information before receiving the 1st request? Is there an easy way to get rid of the need for this dummy publish in every worker/example by adding some initial values in the library or something?

Reply:
The metrics is also used for liveness check that the KV router will use the list of returned metrics to know what are the candidates, so there must be one publish when the worker is ready for requests, to set valid (can be dummy) metrics entry to "be alive".

You can wrap this behavior to be explicit like a "mark-ready" API and user can optionally provide value that actually reflect the current state. But I do want to make this an explicit call as the user may construct the metrics publisher at arbitrary state of the model initialization and the model will not be ready at the point of construction.

Reply:
Thanks for the clarification. I think a "mark-ready" API makes more sense than a dummy call to publish as a reader seeing this for the first time - they may overthink the settings of these metrics values on the dummy call and question their meaning. Can be a separate enhancement.